File: src/dataset/python/say_exercism-ThomasZumsteg/openAI/ChatGPT4_say_v4.py
NUMBERS = ['', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine']
TENS = ['', 'ten', 'twenty', 'thirty', 'forty', 'fifty', 'sixty', 'seventy', 'eighty', 'ninety']
TEENS = ['ten', 'eleven', 'twelve', 'thirteen', 'fourteen', 'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen']
POWERS = ['', ' thousand', ' million', ' billion']

def say(num):
    if not (0 <= num < 1_000_000_000_000):
        raise ValueError("Number must be between 0 and one trillion")
    if num == 0:
        return 'zero'

    parts = []
    for p, power in enumerate(iter_group(num, 1000)):
        if power:
            part = say_power(power) + POWERS[p]
            parts.append(part if p == 0 or power >= 100 else 'and ' + part)
    return ' '.join(reversed(parts))

def iter_group(num, base):
    while num > 0:
        yield num % base
        num //= base

def say_power(num):
    hundreds, remainder = divmod(num, 100)
    tens, ones = divmod(remainder, 10)
    parts = []
    if hundreds:
        parts.append(f"{NUMBERS[hundreds]} hundred")
    if tens == 1:
        parts.append(TEENS[ones])
    else:
        if tens:
            parts.append(TENS[tens])
        if ones:
            parts.append(NUMBERS[ones])
    return ' and '.join(parts) if hundreds and remainder else ' '.join(parts)

File: src/dataset/javascript/say_exercism-ThomasZumsteg/openAI/ChatGPT4_say_v4.js
const ones = ["", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];
const teens = ["", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"];
const tens = ["", "", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"];
const powers = ["", " thousand", " million", " billion"];

function inEnglish(num) {
    if (num < 0 || num >= 1e12) throw new Error('Number must be between 0 and 999,999,999,999.');
    if (num === 0) return "zero";

    const digitGroups = [];
    while (num > 0) {
        digitGroups.push(num % 1000);
        num = Math.floor(num / 1000);
    }

    return digitGroups
        .map((group, power) => group ? englishPower(group) + powers[power] : "")
        .filter(Boolean)
        .reverse()
        .join(" ");
}

function englishPower(digits) {
    const [hundreds, tensDigit, onesDigit] = ("00" + digits).slice(-3).split('').map(Number);
    let english = hundreds ? ones[hundreds] + " hundred" : "";

    if (tensDigit === 1 && onesDigit !== 0) {
        return english + (english ? " " : "") + teens[onesDigit];
    }

    if (tensDigit > 1) {
        english += (english ? " " : "") + tens[tensDigit];
    }

    if (onesDigit > 0) {
        english += (english && tensDigit > 1 ? "-" : english ? " " : "") + ones[onesDigit];
    }

    return english;
}

module.exports = { inEnglish };